api/src/middleware/authentication.ts

```diff
@@ -5,36 +5,47 @@ import * as response from 'lib/response';
 import * as userService from 'user/service';
 import * as authorizationService from 'authorization/service';
 
-const authentication = (optional = false, requiresName = true): middy.MiddlewareObj => {
+const authentication = (optional = false, requiresName = true, endpointSpecificKey?: string): middy.MiddlewareObj => {
     const before: middy.MiddlewareFn<I.APIGatewayProxyEventV2> = async (request): Promise<I.JSONResponse | void> => {
         try {
             let user: null | I.User = null;
 
             const apiKey = request.event.queryStringParameters?.apiKey;
-            const bearerToken = request.event.headers.Authorization;
 
-            if (apiKey) {
-                user = await userService.getByApiKey(apiKey);
-            } else if (bearerToken) {
-                user = authorizationService.validateJWT(bearerToken.split(' ')[1]);
-            }
-
-            if (!optional) {
-                // If there's no user account, and authentication is *not* optional, then the request is blocked.
-                if (!user) {
-                    return response.json(401, { message: 'Please enter either a valid apiKey or bearer token.' });
+            if (endpointSpecificKey) {
+                // The function only accepts one specific api key.)
+                if (apiKey !== endpointSpecificKey) {
+                    return response.json(401, { message: "Please provide a valid 'apiKey'." });
+                }
+            } else {
+                // The function attempts to authenticate an existing user,
+                // looking them up either by their api key or JWT,
+                // and then adds the user details to the request event.
+                const bearerToken = request.event.headers.Authorization;
+
+                if (apiKey) {
+                    user = await userService.getByApiKey(apiKey);
+                } else if (bearerToken) {
+                    user = authorizationService.validateJWT(bearerToken.split(' ')[1]);
                 }
 
-                // If the user hasn't made their name visible in ORCiD, we want to disallow them from doing most actions.
-                if (!user?.firstName && !user?.lastName && requiresName) {
-                    return response.json(403, {
-                        message:
-                            'No name detected. Please ensure your name visibility is set to "Everyone" or "Trusted parties" on your ORCiD account, then re-authorize at /authorization.'
-                    });
+                if (!optional) {
```

Comment:
Could move all this optional stuff up a bit?

```
if (optional) {
    return;
}

 const bearerToken = request.event.headers.Authorization;
...
```

Saves doing all the checks on the api key/jwt only to disregard it because auth is optional.

Reply:
I want to leave the possibility that an endpoint doesn't require a valid authenticated user, but if there is one, the behaviour differs depending on it or uses their details in some way. I think this is done in some places.

Anyway, the early return can still be put in if there is no key/token so I have done that.